# ==============================================================================
# REGRESSOR COMMIT DIFF
# ==============================================================================
# File: js/src/gc/WeakMap.h
# Commit: f7b553bd9c0b
# Full Hash: f7b553bd9c0b77b0130d377f9a9c696c8c611a01
# Author: Steve Fink <sfink@mozilla.com>
# Date: 2021-05-11 21:39:06
# Regressor Bug: 1694538
# File Overlap Count: 1
# Description:
#   Bug 1694538 - Convert gcWeakKeys from <weakmap, key> tuples that require an extra lookup, to <ptr to color, target> tuples that simplify the code substantially r=jonco
#   
#   Differential Revision: https://phabricator.services.mozilla.com/D106226
# ==============================================================================

diff -r 89f7028049ed -r f7b553bd9c0b js/src/gc/WeakMap.h
--- a/js/src/gc/WeakMap.h	Tue May 11 14:01:20 2021 +0000
+++ b/js/src/gc/WeakMap.h	Tue May 11 14:18:38 2021 +0000
@@ -77,10 +77,10 @@
 //   key marked, then map marked:
 //    - value was marked with map in `markEntries()`
 //   map marked, key already in map, key marked before weak marking mode:
-//    - key added to weakKeys when map marked in `markEntries()`
+//    - key added to gcEphemeronEdges when map marked in `markEntries()`
 //    - value marked during `enterWeakMarkingMode`
 //   map marked, key already in map, key marked after weak marking mode:
-//    - when key is marked, weakKeys[key] triggers marking of value in
+//    - when key is marked, gcEphemeronEdges[key] triggers marking of value in
 //      `markImplicitEdges()`
 //   map marked, key inserted into map, key marked:
 //    - value was live when inserted and must get marked at some point
@@ -152,24 +152,14 @@
 
   // We have a key that, if it or its delegate is marked, may lead to a WeakMap
   // value getting marked. Insert it or its delegate (if any) into the
-  // appropriate zone's gcWeakKeys or gcNurseryWeakKeys.
-  static inline void addWeakEntry(GCMarker* marker, gc::Cell* key,
-                                  const gc::WeakMarkable& markable);
+  // appropriate zone's gcEphemeronEdges or gcNurseryEphemeronEdges.
+  inline bool addImplicitEdges(gc::Cell* key, gc::Cell* delegate,
+                               gc::Cell* value);
 
   // Any weakmap key types that want to participate in the non-iterative
   // ephemeron marking must override this method.
   virtual void markKey(GCMarker* marker, gc::Cell* markedCell, gc::Cell* l) = 0;
 
-  // An unmarked CCW with a delegate will add a weakKeys entry for the
-  // delegate. If the delegate is removed with NukeCrossCompartmentWrapper,
-  // then the (former) CCW needs to be added to weakKeys instead.
-  virtual void postSeverDelegate(GCMarker* marker, JSObject* key) = 0;
-
-  // When a wrapper is remapped, it will have its delegate removed then
-  // re-added. Update the delegate zone's gcWeakKeys accordingly.
-  virtual void postRestoreDelegate(GCMarker* marker, JSObject* key,
-                                   JSObject* delegate) = 0;
-
   virtual bool markEntries(GCMarker* marker) = 0;
 
 #ifdef JS_GC_ZEAL
@@ -253,22 +243,6 @@
     return p;
   }
 
-  void remove(Ptr p) {
-    MOZ_ASSERT(p.found());
-    if (mapColor) {
-      forgetKey(p->key());
-    }
-    Base::remove(p);
-  }
-
-  void remove(const Lookup& l) {
-    if (Ptr p = lookup(l)) {
-      remove(p);
-    }
-  }
-
-  void clear();
-
   template <typename KeyInput, typename ValueInput>
   [[nodiscard]] bool add(AddPtr& p, KeyInput&& k, ValueInput&& v) {
     MOZ_ASSERT(k);
@@ -313,18 +287,9 @@
 
   bool markEntry(GCMarker* marker, Key& key, Value& value);
 
-  // 'key' has lost its delegate, update our weak key state.
-  void postSeverDelegate(GCMarker* marker, JSObject* key) override;
-
-  // 'key' regained its delegate, update our weak key state.
-  void postRestoreDelegate(GCMarker* marker, JSObject* key,
-                           JSObject* delegate) override;
-
   void trace(JSTracer* trc) override;
 
  protected:
-  inline void forgetKey(UnbarrieredKey key);
-
   inline void assertMapIsSameZoneWithValue(const Value& v);
 
   bool markEntries(GCMarker* marker) override;